node/buildutils.js

```diff
@@ -57,59 +61,86 @@ exports.getExternals = function () {
     }
 }
 
-var downloadFile = function (url) {
-    // validate parameters
-    if (!url) {
-        throw new Error('Parameter "url" must be set.');
-    }
+exports.getExternalsAsync = getExternalsAsync
 
-    // short-circuit if already downloaded
-    var scrubbedUrl = url.replace(/[/\:?]/g, '_');
-    var targetPath = path.join(downloadPath, 'file', scrubbedUrl);
-    var marker = targetPath + '.completed';
-    if (!test('-f', marker)) {
-        console.log('Downloading file: ' + url);
 
+
+/**
+ * @deprecated This method uses library which is not prefered to use on production
+ */
+
+exports.getExternals = function () {
+    var result = false;
+    getExternalsAsync().then(t => result = true);
+    deasync.loopWhile(function () { return !result });
+    return result;
+}
+
+
+var downloadFileAsync = async function (url, fileName) {
+    return new Promise(async (resolve, reject) => {
+        // validate parameters
+        if (!url) {
+            reject(new Error('Parameter "url" must be set.'));
+        }
+
+        // skip if already downloaded
+        var scrubbedUrl = url.replace(/[/\:?]/g, '_');
+        if (fileName == undefined)
+            fileName = scrubbedUrl
+        var targetPath = path.join(downloadPath, 'file', fileName);
+        var marker = targetPath + '.completed';
+        if (test('-f', marker)) {
+            console.log('File already exist: ' + targetPath);
+            resolve(targetPath)
+            return;
+        }
+
+        console.log('Downloading file: ' + url);
         // delete any previous partial attempt
         if (test('-f', targetPath)) {
             rm('-f', targetPath);
         }
 
         // download the file
         mkdir('-p', path.join(downloadPath, 'file'));
-        var result = syncRequest('GET', url);
-        fs.writeFileSync(targetPath, result.getBody());
 
-        // write the completed marker
-        fs.writeFileSync(marker, '');
-    }
+        const downloader = new Downloader({
+            url: url,
+            directory: path.join(downloadPath, 'file'),
+            fileName: fileName
+        });
 
-    return targetPath;
+        try {
+            const { fileName } = await downloader.download(); //Downloader.download() resolves with some useful properties.
+            fs.writeFileSync(marker, '');
+            resolve(fileName)
+        } catch (error) {
+            reject(error)
+        }
+    });
 }
 
-var downloadArchive = function (url) {
-    // validate platform
-    var platform = os.platform();
-    if (platform != 'darwin' && platform != 'linux') {
-        throw new Error('Unexpected platform: ' + platform);
-    }
-
-    // validate parameters
-    if (!url) {
-        throw new Error('Parameter "url" must be set.');
-    }
 
-    if (!url.match(/\.tar\.gz$/)) {
-        throw new Error('Expected .tar.gz');
-    }
+var downloadArchiveAsync = async function (url, fileName) {
+    return new Promise(async (resolve, reject) => {
```

Comment:
probably we can use just async/await here

Reply:
Fixed